docs/dev-proc/smbios-rules.md: add SMBIOS conventions doc (#64)
Dasharo/docs: docs/dev-proc/smbios-rules.md

# Dasharo Product Guidelines: SMBIOS Information

## Introduction

This document describes the guidelines for filling in the SMBIOS fields in
devices supported by Dasharo. It is meant to be used primarily as a reference
for Dasharo developers.

## BIOS Information (Type 0)

### Vendor field

The vendor field is always set as follows:

```

> 3mdeb Embedded Systems Consulting

Reviewer: ```suggestion
3mdeb
```

This is correct name of the company. Maybe we should mentioned, that for firmware before 13 January 2022 this is as above?

Author: Fixed

```

### Version field

The BIOS version string is defined as follows:

```bash
Dasharo ([major_framework]+[minor_framework]) v[version]
```

- `major_framework` refers to the primary framework in which the firmware is
  developed, e.g. `coreboot` or `edk2`.

- `minor_framework` is the secondary framework, a supporting component for the
  primary framework, e.g. a coreboot payload. May not always be used, e.g. if
  `edk2` is used as the primary framework.

> *For TianoCore UEFIPayload, we use the simpler and more recognizable UEFI name.*

Reviewer: Let's link with our respository where TianoCore UEFIPayload coming from.

Author: Added a link

Examples:

```bash
Dasharo (coreboot+UEFI) v1.0.0    - for coreboot-based builds with TianoCore UEFIPayload
Dasharo (coreboot+SeaBIOS) v1.0.0 - for coreboot-based builds with SeaBIOS payload
Dasharo (coreboot+Heads) v1.0.0   - for coreboot-based builds with Heads payload
Dasharo (edk2) v1.0.0             - for plain edk2-based builds
```

> For the version number fields, refer to [Versioning](../versioning).

Reviewer: We should be consistent with above format:

```
- `version` means version according to official [Dasharo Versioning](../versioning) documentation.
```

Author: Fixed

## System Information (Type 1)

### Devices with an existing BIOS implementation

In this case, set all the fields that are relevant (excluding BIOS version and
vendor) to the same values as the stock firmware.

To obtain SMBIOS values from the stock firmware, either:

> - Obtain the logs from the machin from the cloud

Reviewer: I don't get it. From the machine from the cloud?

Author: I'm referring to the logs generated by `fwdump-docker` which we keep on our cloud

Author: Reworded

# Dasharo Product Guidelines: SMBIOS Information

## Introduction

This document describes the guidelines for filling in the SMBIOS fields in
devices supported by Dasharo. It is meant to be used primarily as a reference
for Dasharo developers.

## BIOS Information (Type 0)

### Vendor field

The vendor field is always set as follows:

```
3mdeb
```

> *for firmware first released before January 13 2022:*

Reviewer: ```suggestion
For firmware with initial release date before 13 January 2021:
```

```
3mdeb Embedded Systems Consulting
```

> ### Version field

Reviewer: ```suggestion
### BIOS Version field
```

The BIOS version string is defined as follows:

```bash
Dasharo ([major_framework]+[minor_framework]) v[version]
```

- `major_framework` refers to the primary framework in which the firmware is
  developed, e.g. `coreboot` or `edk2`.

> 

Author: Removed

- `minor_framework` is the secondary framework, a supporting component for the
  primary framework, e.g. a coreboot payload. May not always be used, e.g. if
  `edk2` is used as the primary framework.

- `version` means version according to official [Dasharo Versioning](../versioning) documentation.

*For [TianoCore UEFIPayloadPkg](https://github.com/Dasharo/edk2/tree/workstation/master/UefiPayloadPkg),
we use the simpler and more recognizable UEFI name.*

Examples:

- `Dasharo (coreboot+UEFI) v1.0.0` - for coreboot-based builds with TianoCore UEFIPayload
- `Dasharo (coreboot+SeaBIOS) v1.0.0` - for coreboot-based builds with SeaBIOS payload
- `Dasharo (coreboot+Heads) v1.0.0` - for coreboot-based builds with Heads payload

> - `Dasharo (edk2) v1.0.0` - for plain edk2-based builds

Reviewer: Now I wonder if this shouldn't be UEFI instead of edk2. @macpijan @miczyg1 @krystian-hebel ?

Reviewer 2: We can also have an older edk2 corebootPayloadPkg. The variety here makes it complex. I don't think anything else than edk2 is worth integrating. We prioritize functionality that addresses customer requirements and only edk2 is able to fulfill this need. I would keep UEFI to simplify things. EDK2 is a set of packages and libraries used to create a UEFI firmware. coreboot is a firmware but an incomplete one. so coreboot+STH is meaningful, but edk2 alone isn't. There would have to be edk2 MinPlatform if anything.

tl;dr
I vote for keeping UEFI

Reviewer: @krystian-hebel we already discussed that topic about using generic UEFI name and we agree that UEFI is better marketing term known to everyone. edk2 is unknown term and it will cause confusion. What we can say it mean in this place is UEFI compatibility, so firmware expose those interfaces - we say nothing about implementation we using, because as @miczyg1 said it complicates whole explanation.

Reviewer 3: +1 for UEFI from me in light of the above arguments. We can worry for different UEFI payloads when (if?) we will have such cases. This name should be rather easy to understand. We explain in SBOM which exact projects were used to build the firmware.

Reviewer: @mkopec please replace edk2 with UEFI.

Author: Replaced

> # DMI/SMBIOS filling convention

Author: This document also covers SMBIOS Type 1 information. Updated the title accordingly

2 types of devices may be supported by Dasharo:

> 

Author: I think it's needed for mkdocs to render unordered lists correctly, at least on my machine

> 

Reviewer: OK.

- Those that already have a supported firmware implementation available
- Those that do not

In the first case, a device may have drivers available in Windows Update or
firmware via fwupd/LVFS. If that is the case, we want to leverage this support
in Dasharo too. Since fwupd and Windows Update identify devices by (among other
things) SMBIOS contents, we do this by matching as many SMBIOS fields as

> possible.

Author: Fixed

{25b6ea34-8f52-598e-a27a-31e03014dbe3}   <- Manufacturer + Family + ProductName + ProductSku + BaseboardManufacturer + BaseboardProduct
{dd71301d-2915-5884-af96-498003ea55bb}   <- Manufacturer + Family + ProductName + ProductSku
{8603f686-41b8-5ea0-bbf8-d3e5cfa9c0ac}   <- Manufacturer + Family + ProductName
{17e12139-ddb8-5e43-ad90-74e69a044051}   <- Manufacturer + ProductSku + BaseboardManufacturer + BaseboardProduct
{cd9d0da1-e88d-5d36-aae5-2ba60e8d0ed6}   <- Manufacturer + ProductSku
{99192755-edd8-5f6d-8c46-e7d9e0b8be5f}   <- Manufacturer + ProductName + BaseboardManufacturer + BaseboardProduct
{0f4d8e7f-9419-5817-a8c5-aa54027c1b35}   <- Manufacturer + ProductName
{17e12139-ddb8-5e43-ad90-74e69a044051}   <- Manufacturer + Family + BaseboardManufacturer + BaseboardProduct
{cd9d0da1-e88d-5d36-aae5-2ba60e8d0ed6}   <- Manufacturer + Family
{a127c2d7-5537-5b72-8bc1-9ccd683f18e6}   <- Manufacturer + EnclosureKind
{96005f57-a38f-5b5e-a26a-50d7286811b0}   <- Manufacturer + BaseboardManufacturer + BaseboardProduct
{56bb94fc-caf1-5863-9e85-1e449a0c5a01}   <- Manufacturer
```

In this case, we set all the fields that are relevant (excluding BIOS version

> and vendor) to the same values as the stock firmware.

Reviewer: At the end of document I realized this should be developers documentation, since it is in development section of the website, but as I wrote above then it should be more step by step guide/process of getting correct values for given fields that have to filled. Maybe even section dedicated to coreboot-specific procedures.

This document is more like end-user explanation then developer documentation. Please be clear at the beginning what audience this document is for.

> and vendor) to the same values as the stock firmware.

Author: Added an introduction describing what this document is - guidelines for developers. Restructured the document accordingly

Be aware that sometimes this means a downgrade - in the case above, it would
make more sense to set the `Manufacturer` field to a sensible value such as
`NovaCustom`, but since this field is used in several GUIDs in fwupd and

> Windows Update and the OEM set it to `Notebook`, we'll do this as well.

Author: Changed the document tone to be a developer specification instead of an end-user facing document. Do we still want such a FAQ section?

Exceptions can be made depending on circumstances (e.g. if the field in question

> is important for functionality).

Author: We don't, I thought it may be a good idea to leave the option to change a field in case it's not needed for the specific device. I'll remove this then

> is important for functionality).

Author: Removed

In the case where the device does not have an existing supported BIOS or the
client chooses to use different SMBIOS fields (e.g. to make it easier to

> differentiate devices), we want to set the fields to the most sensible values:

Author: Changed to a more definitive tone

- `Family`: family of devices (e.g. `Clevo_tgl-u` for Clevo Tiger Lake laptops)
- `Manufacturer`: The name of the party selling the device (OEM)
- `ProductName`: The name under which the device is sold
- `BaseboardManufacturer`: The manufacturer of the device (ODM), e.g. `Clevo`
- `BaseboardProduct`: The name of the mainboard part
- `ProductSku`: Variant of the device, e.g. soldered RAM capacity or discrete

>   GPU option. May be read from straps in mainboard code.

Author: Added a link to the SMBIOS 3.5.0 specification and removed reduntant field descriptions

For a list of SMBIOS fields used by Windows Update, see

> [Microsoft's documentation](https://docs.microsoft.com/en-us/windows-hardware/drivers/dashboard/using-chids).

Author: Removed this comment

# BIOS Vendor and Version fields

We use the following convention for filling in the BIOS version fields:

```bash

> Dasharo ([major_framework]+[minor_framework]) v[version]

Author: Added a remark that the `minor_framework` field may not always be used, and added an example to the list

```

`major_framework` refers to the primary framework in which the firmware is
developed, e.g. `coreboot` or `TianoCore`.

`minor_framework` is the secondary framework, a supporting component for the

> primary framework, e.g. a coreboot payload.

Author: Added a note about this

Examples:

```bash
Dasharo (coreboot+UEFI) v1.0.0    - for coreboot-based TianoCore builds
Dasharo (coreboot+SeaBIOS) v1.0.0 - for coreboot-based SeaBIOS builds
Dasharo (coreboot+Heads) v1.0.0   - for coreboot-based Heads builds

> ```

Reviewer: I don't like formatting, better would be:

Examples:
- `Dasharo (coreboot+UEFI) v1.0.0` - for coreboot-based builds with TianoCore payload (where TianoCore payload should be replaced with precise name of component we using)

Author: Fixed

Reviewer: @mkopec not exactly, all text in your case is in code block, in my case only the BIOS Version is in backticks rest is in notmal text.
